### PR TITLE
[SYCL-MLIR] Only transform loops with shared memory accesses

### PR DESCRIPTION
**Pull request by whitneywhtsang**

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 1]
```diff
@@ -1090,10 +1098,10 @@ MemoryAccess::getIntraThreadAccessMatrix(unsigned numGridDimensions) const {
          "Expecting 'numGridDimensions' to be smaller than the number of "
          "columns in the memory access matrix");
   if (numGridDimensions == 0)
-    return matrix;
+    return {};
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:1101]: The definition of intra thread and inter thread was opposite.

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 2]
```diff
@@ -864,8 +864,12 @@ bool MemoryAccessMatrix::hasStridedAccessPattern(DataFlowSolver &solver) const {
         return false;
 
       if (isOnDiagonal) {
+        bool isFirstDiagonalElem = (row == 0 && col == 0);
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:867]: 1 0 0
0 0 0
0 0 0
should also be considered as strided.
This is needed for `affine_3d` and `scf_3d` test cases in `loop_internalization.mlir`.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 3]
```diff
@@ -63,24 +63,17 @@ bool isCandidateFunction(FunctionOpInterface func) {
 /// A loop nest is a candidate iff is perfect and contains only affine or scf
 /// for loops.
 bool isCandidateLoopNest(LoopLikeOpInterface loop) {
-  if (!LoopTools::isOutermostLoop(loop)) {
-    LLVM_DEBUG(llvm::dbgs() << "not candidate: not outermost loop\n");
+  if (!LoopTools::isOutermostLoop(loop))
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:66]: remove debug messages, as they are now misleading after we changed to only print candidate loops.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 4]
```diff
@@ -165,7 +158,7 @@ class MemorySelector {
   DenseMap<Value, SmallVector<affine::MemRefAccess>> accesses;
 
   /// The preferred memory space for each memref access;
-  DenseMap<const affine::MemRefAccess *, MemorySpace> accessToMemSpace;
+  DenseMap<const Operation *, MemorySpace> accessToMemSpace;
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:161]: address of MemRefAccess are different when insert and when query, as they are created locally, need to use the Operation `access.opInst` instead.

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 5]
```diff
@@ -1104,10 +1112,10 @@ MemoryAccess::getInterThreadAccessMatrix(unsigned numGridDimensions) const {
          "Expecting 'numGridDimensions' to be smaller than the number of "
          "columns in the memory access matrix");
   if (numGridDimensions == 0)
-    return {};
+    return matrix;
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:1115]: matrix -> {}

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 6]
```diff
@@ -282,14 +275,27 @@ void MemorySelector::analyze(LoopLikeOpInterface loop, AccessKind accessKind) {
           bool strideIsLargeEnough = stride->sgt(8) || stride->slt(-8);
           useSharedMemory = hasTemporalReuse(access, threadVars) &&
                             (stride->isZero() || strideIsLargeEnough);
-        }
-
-        accessToMemSpace[&access] =
+          // FIXME: getConstIntegerValue doesn't return zero for
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:278]: We have this "recovery code" in another place (isEqual?). Makes me think we could place it in the `getConstIntegerValue` so consumer do not have to all add this code.

**Review by etiotto — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
This is a WIP feature, so merging with one approval. Post-commit review comments are welcome.

**Pull request merged**